File: src/03-password-generator/rust/src/main.rs
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
 */

/*!
 *  This code is based on my Arcanus Psw project.
 *
 *  https://github.com/krondorl/arcanus-psw
 *
 *  Copyright (c) 2023 Adam Burucs. MIT license.
 */

use rand::Rng;

const VOWELS: [&str; 5] = ["a", "e", "i", "o", "u"];
const CONSONANTS: [&str; 21] = [
    "b", "c", "d", "f", "g", "h", "j", "k", "l", "m", "n", "p", "q", "r", "s", "t", "v", "w", "x",
    "y", "z",
];
const NUMBERS: [&str; 10] = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
const SPECIALS: [&str; 6] = ["!", "+", "#", "/", "$", "?"];

fn generate_numbers(length: u8) -> Result<String, String> {
    if (1..=4).contains(&length) {
        let mut generated_numbers: String = Default::default();
        let mut i = 0;
        while i < length {
            let random_number = rand::thread_rng().gen_range(0..9);
            generated_numbers.push_str(NUMBERS[random_number]);
            i += 1;
        }
        Ok(generated_numbers)
    } else {
        Err(String::from(
            "Error by generating words: length parameter should be between 1 and 4.",
        ))
    }
}

fn generate_specials() -> String {
    let random_special = rand::thread_rng().gen_range(0..5);
    SPECIALS[random_special].to_string()
}

fn generate_words(length: u8) -> Result<String, String> {
    if (13..=64).contains(&length) {
        let mut generated_words: String = Default::default();
        let mut i = 0;
        while i < length {
            if i % 2 == 0 {
                let random_consonant = rand::thread_rng().gen_range(0..20);
                if i == 0 || i % 4 == 0 {
                    let uppercase = str::to_uppercase(CONSONANTS[random_consonant]);
                    generated_words.push_str(&uppercase);
                } else {
                    generated_words.push_str(CONSONANTS[random_consonant]);
                }
            } else {
                let random_vowel = rand::thread_rng().gen_range(0..4);
                generated_words.push_str(VOWELS[random_vowel]);
            }
            i += 1;
        }
        Ok(generated_words)
    } else {
        Err(String::from(
            "Error by generating words: length parameter should be between 13 and 64.",
        ))
    }
}

fn generate_password(length: Option<u8>) -> Result<String, String> {
    match length {
        Some(length_value) => {
            if (16..=64).contains(&length_value) {
                let mut generated_password: String = Default::default();
                let words = generate_words(length_value - 3);
                match words {
                    Ok(words_value) => {
                        generated_password.push_str(&words_value);
                        let numbers = generate_numbers(2);
                        match numbers {
                            Ok(numbers_value) => {
                                generated_password.push_str(&numbers_value);
                                let specials = generate_specials();
                                generated_password.push_str(&specials);
                                Ok(generated_password)
                            }
                            Err(e) => Err(e),
                        }
                    }
                    Err(e) => Err(e),
                }
            } else {
                Err(String::from(
                    "Error: generate password should have a length between 16 and 64.",
                ))
            }
        }
        None => {
            let mut generated_password: String = Default::default();
            let words = generate_words(13);
            match words {
                Ok(words_value) => {
                    generated_password.push_str(&words_value);
                    let numbers = generate_numbers(2);
                    match numbers {
                        Ok(numbers_value) => {
                            generated_password.push_str(&numbers_value);
                            let specials = generate_specials();
                            generated_password.push_str(&specials);
                            Ok(generated_password)
                        }
                        Err(e) => Err(e),
                    }
                }
                Err(e) => Err(e),
            }
        }
    }
}

fn main() {
    println!("Password generator");
    let p32 = generate_password(Some(32)).unwrap();
    println!("{:#?}", p32);
}

File: src/03-password-generator/typescript/src/main.ts
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
 */

/*!
 *  This code is based on my Arcanus Psw project.
 *
 *  https://github.com/krondorl/arcanus-psw
 *
 *  Copyright (c) 2023 Adam Burucs. MIT license.
 */

// The node:crypto module provides cryptographically secure random numbers.
import * as crypto from "crypto";

const VOWELS = ["a", "e", "i", "o", "u"] as const;
const CONSONANTS = [
  "b",
  "c",
  "d",
  "f",
  "g",
  "h",
  "j",
  "k",
  "l",
  "m",
  "n",
  "p",
  "q",
  "r",
  "s",
  "t",
  "v",
  "w",
  "x",
  "y",
  "z",
] as const;
const NUMBERS = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"] as const;
const SPECIALS = ["!", "+", "#", "/", "$", "?"] as const;

function generateRandomRange(max: number): number {
  return crypto.randomInt(max);
}

function generateNumbers(length: number): string {
  if (length >= 1 && length <= 4) {
    const generatedNumbers: string[] = [];
    let i = 0;
    while (i < length) {
      let randomNumber = generateRandomRange(9);
      generatedNumbers.push(NUMBERS[randomNumber]);
      i += 1;
    }
    return generatedNumbers.join("");
  } else {
    return "Error by generating words: length parameter should be between 1 and 4.";
  }
}

function generateSpecials(): string {
  const randomRange = generateRandomRange(5);
  return SPECIALS[randomRange];
}

function generateWords(length: number): string {
  if (length >= 13 && length <= 64) {
    const generatedWords: string[] = [];
    let i = 0;
    while (i < length) {
      if (i % 2 === 0) {
        let randomConsonant = generateRandomRange(20);
        if (i === 0 || i % 4 === 0) {
          let uppercase = CONSONANTS[randomConsonant].toUpperCase();
          generatedWords.push(uppercase);
        } else {
          generatedWords.push(CONSONANTS[randomConsonant]);
        }
      } else {
        let randomVowel = generateRandomRange(4);
        generatedWords.push(VOWELS[randomVowel]);
      }
      i += 1;
    }
    return generatedWords.join("");
  } else {
    return "Error by generating words: length parameter should be between 13 and 64.";
  }
}

function generatePassword(length: number): string {
  if (length >= 13 && length <= 64) {
    let generatedPassword: string[] = [];
    let words = generateWords(length - 3);
    if (words.length > 0) {
      generatedPassword.push(words);
      let numbers = generateNumbers(2);
      if (numbers.length > 0) {
        generatedPassword.push(numbers);
        let specials = generateSpecials();
        generatedPassword.push(specials);
        return generatedPassword.join("");
      } else {
        return "Error: wrong numbers length when using generatePassword function.";
      }
    } else {
      return "Error: wrong words length when using generatePassword function.";
    }
  } else {
    return "Error: generate password should have a length between 16 and 64.";
  }
}

function main() {
  console.log("Password generator");
  let p32 = generatePassword(32);
  console.log(p32);
}

main();
